taichi/program/program.h

```diff
@@ -125,8 +126,12 @@ class TI_DLL_EXPORT Program {
 
   CompileConfig &this_thread_config() {
     auto thread_id = std::this_thread::get_id();
+    std::shared_lock<std::shared_mutex> sharedLock(compile_config_mut);
     if (!configs.count(thread_id)) {
+      sharedLock.unlock();
+      std::unique_lock<std::shared_mutex> uniqueLock(compile_config_mut);
       configs[thread_id] = configs[main_thread_id_];
```

Comment:
When you append elements into the hash table, there is a chance of causing rehashing that invalidate all the existing iterators. 

`item #4` of https://en.cppreference.com/w/cpp/container#Thread_safety states that: 
`Container operations that invalidate any iterators modify the container and cannot be executed concurrently with any operations on existing iterators even if those iterators are not invalidated`. So you have to apply a mutex when reading from `configs`, which dramatically reduces the multi-thread performance.

I'd suggest calling `unordered_map::rehash(num_compile_threads)` in main thread beforehand, so as to avoid rehashing. (https://cplusplus.com/reference/unordered_map/unordered_map/rehash/)

Reply:
@lin-hitonami please make sure to add enough context in comment so that it's easier for readers to understand why the rehash is important. ;)